File: src/System/Mesos/Raw/OfferId.hs
module System.Mesos.Raw.OfferId where
import           System.Mesos.Internal

type OfferIDPtr = Ptr OfferID

foreign import ccall unsafe "ext/types.h toOfferID" c_toOfferID :: ToID OfferIDPtr

foreign import ccall unsafe "ext/types.h fromOfferID" c_fromOfferID :: FromID OfferIDPtr

foreign import ccall unsafe "ext/types.h destroyOfferID" c_destroyOfferID :: OfferIDPtr -> IO ()

instance CPPValue OfferID where

  marshal x = do
    (strp, l) <- cstring $ fromOfferID x
    liftIO $ c_toOfferID strp $ fromIntegral l

  unmarshal p = fmap OfferID $ do
    ptrPtr <- alloc
    len <- liftIO $ c_fromOfferID p ptrPtr
    peekCString' (ptrPtr, len)

  destroy = c_destroyOfferID

File: ext/types/offer_id.cpp
#include <iostream>
#include "types.h"

using namespace mesos;

OfferIDPtr toOfferID(char* bs, int len)
{
  OfferIDPtr val = new OfferID();
  val->set_value(bs, len);
  return val;
}

int fromOfferID(OfferIDPtr p, char** poke)
{
  *poke = (char*) p->mutable_value()->data();
  return p->value().size();
}

void destroyOfferID(OfferIDPtr p)
{
  delete p;
}
